rafthttp/stream.go

```diff
@@ -0,0 +1,388 @@
+		resp.Body.Close()
+		return nil, fmt.Errorf("unhandled http status %d", resp.StatusCode)
+	}
+
+	stream := &incomingStream{}
+	switch d.t {
+	case streamTypeMsgApp:
+		stream.decoder = &msgAppDecoder{r: resp.Body, local: d.from, remote: d.to, term: term}
+	case streamTypeGeneral:
+		stream.decoder = &generalDecoder{r: resp.Body}
+	default:
+		log.Panicf("rafthttp: unhandled stream type %s", d.t)
+	}
+	stream.Closer = resp.Body
+	return stream, nil
+}
+
+func (d *streamDialer) update(u string) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.u = u
+	select {
+	case d.updatec <- struct{}{}:
+	default:
+	}
+}
+
+func (d *streamDialer) updateMsgAppTerm(term uint64) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.msgAppTerm == term {
+		return
+	}
+	d.msgAppTerm = term
+	select {
+	case d.updatec <- struct{}{}:
+	default:
+	}
+}
+
+type streamReader struct {
+	dialer *streamDialer
+	recvc  chan<- raftpb.Message
+	stopc  chan struct{}
+	done   chan struct{}
+}
+
+func startStreamReader(d *streamDialer, recvc chan<- raftpb.Message) *streamReader {
+	r := &streamReader{
+		dialer: d,
+		recvc:  recvc,
+		stopc:  make(chan struct{}),
+		done:   make(chan struct{}),
+	}
+	go r.run()
+	return r
+}
+
+func (cr *streamReader) run() {
+	var stream *incomingStream
+	var err error
+	reset := func() {
+		if stream != nil {
+			stream.Close()
+			stream = nil
+		}
+	}
+	for {
+		select {
+		case <-cr.dialer.updatec:
+			reset()
+		case <-cr.stopc:
+			reset()
+			close(cr.done)
+			return
+		default:
+		}
+		if stream == nil {
+			stream, err = cr.dialer.dial()
+			if err != nil {
+				log.Printf("rafthttp: dial error: %v", err)
+				continue
+			}
+		}
+		m, err := stream.decode()
+		switch {
+		case err != nil && err != io.EOF:
+			log.Printf("rafthttp: encountered error reading the connection: %v", err)
+			fallthrough
+		case err != nil:
+			reset()
+		case isHeartbeatMessage(m):
+		default:
+			select {
+			case cr.recvc <- m:
+			default:
+				log.Printf("rafthttp: dropping %s from %s because receive buffer is blocked",
+					m.Type, m.From)
+			}
+		}
+	}
+}
+
+func (cr *streamReader) stop() {
+	close(cr.stopc)
+	<-cr.done
+}
+
+type writeFlushCloser interface {
+	io.WriteCloser
+	http.Flusher
+}
+
+type outgoingConn struct {
+	t streamType
+	writeFlushCloser
+	info http.Header
+}
+
+type streamPicker struct {
+	fs *stats.FollowerStats
+
+	mu           sync.Mutex // protect msgAppTerm
+	msgAppTerm   uint64
+	msgAppWriter *streamWriter
+	writer       *streamWriter
+}
+
+func newStreamPicker(fs *stats.FollowerStats) *streamPicker {
+	return &streamPicker{
+		fs:           fs,
+		msgAppWriter: startStreamWriter(),
+		writer:       startStreamWriter(),
+	}
+}
+
+func (p *streamPicker) pick(m raftpb.Message) *streamWriter {
+	switch {
+	case p.canUseMsgAppStream(m) && p.msgAppWriter.isWorking():
+		return p.msgAppWriter
+	case p.writer.isWorking():
+		return p.writer
+	default:
+		return nil
+	}
+}
+
+func (p *streamPicker) attachOutgoingConn(conn outgoingConn) {
```

Comment:
Both are fine. I prefer this one because it is easy to think that the function takes the ownership of connection from now on.